woocommerce/woocommerce-gutenberg-products-block
# Add wp.data store for schema and lists.

This pull adds all the initial data stores needed for implementing the [All Products](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/899) block.  They also will be the primary resource for future interfaces with checkout/cart work and eventual refactoring of existing blocks away from direct api requests.

In this pull there are three stores introduced:

* [x] `schema`: primarily for getting schema details via rest request and exposing that for other stores to use.  Initially, there was just focus on getting route info for a given namespace via this store.  This store should rarely be used externally as it's mostly used for other stores to get routes for a request and (eventually) validate using the schema provided via route apis.
* [x] `collections`:  a simple cached store of query strings -> results. 
* [x] `query-state`: a store for caching active filter queries.  This will be used for any sharing of query keys and their values across a given context (i.e. block or group of blocks) typically used in queries for those blocks (via the `collections` store).

For easier review, I rebased the work into grouped commits:

- 3a4b912 adds the `@wordpress/data-controls` and `deep-freeze` dependencies (`deepFreeze` is useful for tests).
- 658c23e adds the code for the `schema` store.
- a0e45ba adds the code for the `query-state` store.
- 008234d adds the code for the `collections` store.
- 7b23f84 adds the code for the store bundle entry point and exports.
- 12609b9 adds the main `README.md` file for the new `wc.wcBlocksData` bundle.
- f7b8512 adds the webpack configuration and php side registration for the new data store bundle.

## Basic architecture considerations:

- The stores are all exported to a bundle (`wc-blocks-data.js`) and registered when loaded.  The store keys are exported on the `@woocommerce/blocks-data` external (see [doc](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/f7b85123df22f94291f4656d1ebb623b9e7cf063/assets/js/data/README.md))
- The stores are all registered as a unit when `wc-blocks-data-store` is registered as a dependency handle.
- the `collections` store is only for read (`GET`) requests only.  Future needs may require other types of requests and we can introduce specific new stores for those as necessary.

You can read more about the implementation/usage of the stores via their internal docs:

- `collections`: [doc](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/f7b85123df22f94291f4656d1ebb623b9e7cf063/assets/js/data/collections/README.md)
- `schema`: [doc](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/f7b85123df22f94291f4656d1ebb623b9e7cf063/assets/js/data/schema/README.md)
- `query-state`: [doc](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/f7b85123df22f94291f4656d1ebb623b9e7cf063/assets/js/data/query-state/README.md)

## To Test:

- The stores do have coverage with unit tests

I also did some monkey testing by adding `wc-blocks-data-store` as a dependency on `wc-blocks`.  That way when loading the block-editor I can play around with the stores via the console.  For example, executing this should return an array of products:

```js
storeKey = wc.wcBlocksData.COLLECTIONS_STORE_KEY;
wp.data.select( storeKey ).getCollection( '/wc/blocks', 'products' );
// might have to execute more than once in the console because this has a resolver, so it needs to resolve... normally this is done in a subscription callback.
wp.data.select( storeKey ).getCollection( '/wc/blocks', 'products' );
```

## Next Steps:

- This pull is based off the `add/all-products-block` branch.  So it would get merged back into that.
- Will need to wire up the `withProducts` hoc to use data stores instead of direct api requests. 
- part of wiring up this to the All Products block will be implementing the back-compat build requirements (and the dependency on WP 5.3 for the block).
- Will wire up the `filter by price` block so that any changes to query state change the query fed into the `collections` select.

... in the future
- consider contexts for endpoints (i.e. edit vs public) which may require some work on the api overrides we add for various woo models. Contexts can be automatically injected using a custom wp.apiFetch middleware.
- since this will implement reading the schema from the api endpoint, we can consider (likely not this pull) implementing an abstraction for query params so validation etc can happen automatically against the schema (i.e. validate query arguments passed into `collections` selectors to ensure they are valid for the given route)
